## Упражнения по библиотеке Numpy

In [1]:
import numpy as np


**1.** Дан случайный массив, поменять знак у элементов, значения которых между 3 и 8

In [2]:
arr = np.random.uniform(0, 10, 10)
print("Original array:\n", arr)

arr[(arr >= 3) & (arr <= 8)] *= -1

print("\nModified array:\n", arr)

Original array:
 [3.79618652 2.10385175 0.61709474 3.20438302 4.44124273 5.87419691
 3.90367053 3.1225819  0.05339208 7.61616792]

Modified array:
 [-3.79618652  2.10385175  0.61709474 -3.20438302 -4.44124273 -5.87419691
 -3.90367053 -3.1225819   0.05339208 -7.61616792]


**2.** Заменить максимальный элемент случайного массива на 0

In [3]:
arr = np.random.randint(1, 21, size=5)
print("Original array:", arr)

arr[arr == np.max(arr)] = 0


print("Modified array:", arr)


Original array: [16  3 13 12  5]
Modified array: [ 0  3 13 12  5]


**3.** Построить прямое произведение массивов (все комбинации с каждым элементом). На вход подается двумерный массив

In [4]:
A = np.array([[1, 2], 
              [3, 4]])

B = np.array([[5, 6], 
              [7, 8]])

result = np.multiply.outer(A, B)

print("Resulting Shape:", result.shape)
print("Resulting 4D Array:\n", result)


Resulting Shape: (2, 2, 2, 2)
Resulting 4D Array:
 [[[[ 5  6]
   [ 7  8]]

  [[10 12]
   [14 16]]]


 [[[15 18]
   [21 24]]

  [[20 24]
   [28 32]]]]


**4.** Даны 2 массива A (8x3) и B (2x2). Найти строки в A, которые содержат элементы из каждой строки в B, независимо от порядка элементов в B

In [5]:
A = np.array([
    [1, 5, 3],  
    [1, 2, 5], 
    [7, 8, 9],
    [2, 4, 10],
    [3, 4, 1],
    [0, 0, 0],
    [1, 1, 1],
    [2, 3, 2]
])

B = np.array([
    [1, 2],    
    [3, 4]
])

matches = A[:, None, :, None] == B[None, :, None, :]

row_has_match_with_B_row = matches.any(axis=(2, 3))

final_mask = row_has_match_with_B_row.all(axis=1)

matching_rows = A[final_mask]

print("Matching Rows:\n", matching_rows)


Matching Rows:
 [[ 1  5  3]
 [ 2  4 10]
 [ 3  4  1]
 [ 2  3  2]]


**5.** Дана 10x3 матрица, найти строки из неравных значений (например строка [2,2,3] остается, строка [3,3,3] удаляется)

In [6]:
matrix = np.array([
    [2, 2, 3], 
    [3, 3, 3],
    [4, 5, 6],
    [1, 1, 1],  
    [7, 8, 9],
    [5, 5, 6],
    [9, 9, 9],
    [0, 2, 4],
    [8, 8, 1],
    [6, 6, 6]
])

all_equal_mask = np.all(matrix == matrix[:, [0]], axis=1)

unequal_rows = matrix[~all_equal_mask]

print("Rows with unequal values:\n", unequal_rows)


Rows with unequal values:
 [[2 2 3]
 [4 5 6]
 [7 8 9]
 [5 5 6]
 [0 2 4]
 [8 8 1]]


**6.** Дан двумерный массив. Удалить те строки, которые повторяются

In [7]:
matrix = np.array([
    [2, 2, 3], 
    [6, 6, 6],
    [4, 5, 6],
    [1, 1, 1],  
    [7, 8, 9],
    [6, 6, 6],
    [7, 8, 9],
    [6, 6, 6],
])

unique_rows = np.unique(matrix, axis=0)

print("Original:\n", matrix)
print("\nMatrix after deleting duplicates:\n", unique_rows)


Original:
 [[2 2 3]
 [6 6 6]
 [4 5 6]
 [1 1 1]
 [7 8 9]
 [6 6 6]
 [7 8 9]
 [6 6 6]]

Matrix after deleting duplicates:
 [[1 1 1]
 [2 2 3]
 [4 5 6]
 [6 6 6]
 [7 8 9]]


______
______

Для каждой из следующих задач (1-5) нужно привести 2 реализации – одна без использования numpy (cчитайте, что там, где на входе или выходе должны быть numpy array, будут просто списки), а вторая полностью векторизованная с использованием numpy (без использования питоновских циклов/map/list comprehension).


__Замечание 1.__ Можно считать, что все указанные объекты непустые (к примеру, в __задаче 1__ на диагонали матрицы есть ненулевые элементы).

__Замечание 2.__ Для большинства задач решение занимает не больше 1-2 строк.

___

* __Задача 1__: Подсчитать произведение ненулевых элементов на диагонали прямоугольной матрицы.  
 Например, для X = np.array([[1, 0, 1], [2, 0, 2], [3, 0, 3], [4, 4, 4]]) ответ 3.

In [8]:
def diag_prod_no_numpy(X):
    min_dim = min(len(X), len(X[0]))
    product = 1
    for i in range(min_dim):
        val = X[i][i]
        if val != 0:
            product *= val
    return product

X_list = [[1, 0, 1], [2, 0, 2], [3, 0, 3], [4, 4, 4]]
diag_prod_no_numpy(X_list)

3

In [9]:
X = np.array([[1, 0, 1], [2, 0, 2], [3, 0, 3], [4, 4, 4]])
diag = X.diagonal()
np.prod(diag[diag != 0])

3

* __Задача 2__: Даны два вектора x и y. Проверить, задают ли они одно и то же мультимножество.  
  Например, для x = np.array([1, 2, 2, 4]), y = np.array([4, 2, 1, 2]) ответ True.

In [10]:
def is_same_multiset_no_numpy(x, y):
    if len(x) != len(y):
        return False
    return sorted(x) == sorted(y)

x_list = [1, 2, 2, 4]
y_list = [4, 2, 1, 2]
is_same_multiset_no_numpy(x_list, y_list)


True

In [11]:
x = np.array([1, 2, 2, 4])
y = np.array([4, 2, 1, 2])

np.array_equal(np.sort(x), np.sort(y))

True

* __Задача 3__: Найти максимальный элемент в векторе x среди элементов, перед которыми стоит ноль. 
 Например, для x = np.array([6, 2, 0, 3, 0, 0, 5, 7, 0]) ответ 5.

In [12]:
def max_after_zero_no_numpy(x):
    max_val = float('-inf')
    for i in range(len(x) - 1):
        if x[i] == 0:
            max_val = max(max_val, x[i+1])
    return max_val

x_list = [6, 2, 0, 3, 0, 0, 5, 7, 0]
max_after_zero_no_numpy(x_list)


5

In [13]:
x = np.array(x_list)

mask = (x[:-1] == 0)

np.max(x[1:][mask])

5

* __Задача 4__: Реализовать кодирование длин серий (Run-length encoding). Для некоторого вектора x необходимо вернуть кортеж из двух векторов одинаковой длины. Первый содержит числа, а второй - сколько раз их нужно повторить.  
 Например, для x = np.array([2, 2, 2, 3, 3, 3, 5]) ответ (np.array([2, 3, 5]), np.array([3, 3, 1])).

In [14]:
def rle_no_numpy(x):
    if not x:
        return ([], [])
        
    values = [x[0]]
    counts = [1]
    
    for i in range(1, len(x)):
        if x[i] == x[i-1]:
            counts[-1] += 1
        else:
            values.append(x[i])
            counts.append(1)
            
    return (values, counts)

x_list = [2, 2, 2, 3, 3, 3, 5]
rle_no_numpy(x_list)

([2, 3, 5], [3, 3, 1])

In [15]:
x  = np.array([2, 2, 2, 3, 3, 3, 5])

indices = np.nonzero(x[:-1] != x[1:])[0]

idx = np.concatenate(([-1], indices, [len(x) - 1]))

values = x[idx[1:]]
counts = np.diff(idx)

(values, counts) = (values, counts)
print((values, counts))

(array([2, 3, 5]), array([3, 3, 1]))


* __Задача 5__: Даны две выборки объектов - X и Y. Вычислить матрицу евклидовых расстояний между объектами. Сравните с функцией scipy.spatial.distance.cdist по скорости работы.

In [16]:

import math
def cdist_no_numpy(X, Y):
    num_X = len(X)
    num_Y = len(Y)
    dim = len(X[0])
    
    dist_matrix = [[0.0 for _ in range(num_Y)] for _ in range(num_X)]
    
    for i in range(num_X):
        for j in range(num_Y):
            squared_sum = 0.0
            for d in range(dim):
                squared_sum += (X[i][d] - Y[j][d]) ** 2
            dist_matrix[i][j] = math.sqrt(squared_sum)
            
    return dist_matrix

X_list = [[1.0, 2.0], [3.0, 4.0]]
Y_list = [[5.0, 6.0], [1.0, 2.0], [0.0, 0.0]]
cdist_no_numpy(X_list, Y_list)

[[5.656854249492381, 0.0, 2.23606797749979],
 [2.8284271247461903, 2.8284271247461903, 5.0]]

In [17]:
def cdist_numpy(X, Y):
    return np.sqrt(np.sum((X[:, None, :] - Y[None, :, :]) ** 2, axis=-1))

X_np = np.array([[1.0, 2.0], [3.0, 4.0]])
Y_np = np.array([[5.0, 6.0], [1.0, 2.0], [0.0, 0.0]])
cdist_numpy(X_np, Y_np)


array([[5.65685425, 0.        , 2.23606798],
       [2.82842712, 2.82842712, 5.        ]])

_______
________

* #### __Задача 6__: CrunchieMunchies __*__

Вы работаете в отделе маркетинга пищевой компании MyCrunch, которая разрабатывает новый вид вкусных, полезных злаков под названием **CrunchieMunchies**.

Вы хотите продемонстрировать потребителям, насколько полезны ваши хлопья по сравнению с другими ведущими брендами, поэтому вы собрали данные о питании нескольких разных конкурентов.

Ваша задача - использовать вычисления Numpy для анализа этих данных и доказать, что ваши **СrunchieMunchies** - самый здоровый выбор для потребителей.


In [18]:
import numpy as np

1. Просмотрите файл cereal.csv. Этот файл содержит количества калорий для различных марок хлопьев. Загрузите данные из файла и сохраните их как calorie_stats.

In [19]:
calorie_stats = np.loadtxt("./data/cereal.csv", delimiter=",")
calorie_stats

array([ 70., 120.,  70.,  50., 110., 110., 110., 130.,  90.,  90., 120.,
       110., 120., 110., 110., 110., 100., 110., 110., 110., 100., 110.,
       100., 100., 110., 110., 100., 120., 120., 110., 100., 110., 100.,
       110., 120., 120., 110., 110., 110., 140., 110., 100., 110., 100.,
       150., 150., 160., 100., 120., 140.,  90., 130., 120., 100.,  50.,
        50., 100., 100., 120., 100.,  90., 110., 110.,  80.,  90.,  90.,
       110., 110.,  90., 110., 140., 100., 110., 110., 100., 100., 110.])

2. В одной порции CrunchieMunchies содержится 60 калорий. Насколько выше среднее количество калорий у ваших конкурентов?

Сохраните ответ в переменной average_calories и распечатайте переменную в терминале

In [20]:
average_calories = np.mean(calorie_stats)

print(f"Competitors' average: {average_calories:.2f} kcal")
print(f"Higher CrunchieMunchies by: {average_calories - 60:.2f} kcal")

Competitors' average: 106.88 kcal
Higher CrunchieMunchies by: 46.88 kcal


3. Корректно ли среднее количество калорий отражает распределение набора данных? Давайте отсортируем данные и посмотрим.

Отсортируйте данные и сохраните результат в переменной calorie_stats_sorted. Распечатайте отсортированную информацию

In [21]:
calorie_stats_sorted = np.sort(calorie_stats)
print(calorie_stats_sorted)

[ 50.  50.  50.  70.  70.  80.  90.  90.  90.  90.  90.  90.  90. 100.
 100. 100. 100. 100. 100. 100. 100. 100. 100. 100. 100. 100. 100. 100.
 100. 100. 110. 110. 110. 110. 110. 110. 110. 110. 110. 110. 110. 110.
 110. 110. 110. 110. 110. 110. 110. 110. 110. 110. 110. 110. 110. 110.
 110. 110. 110. 120. 120. 120. 120. 120. 120. 120. 120. 120. 120. 130.
 130. 140. 140. 140. 150. 150. 160.]


4. Похоже, что большинство значений выше среднего. Давайте посмотрим, является ли медиана наиболее корректным показателем набора данных.

Вычислите медиану набора данных и сохраните свой ответ в median_calories. Выведите медиану, чтобы вы могли видеть, как она сравнивается со средним значением.

In [22]:
median_calories = np.median(calorie_stats)

print(f"Median: {median_calories:.1f} kca;")
print(f"Mean: {average_calories:.2f} kcal")

Median: 110.0 kca;
Mean: 106.88 kcal


5. В то время как медиана показывает, что по крайней мере половина наших значений составляет более 100 калорий, было бы более впечатляюще показать, что значительная часть конкурентов имеет более высокое количество калорий, чем CrunchieMunchies.

Рассчитайте различные процентили и распечатайте их, пока не найдете наименьший процентиль, превышающий 60 калорий. Сохраните это значение в переменной nth_percentile.

In [23]:
nth_percentile = None

for p in range(1, 101):
    value = np.percentile(calorie_stats, p)
    print(f"{p:>3}-th percentile: {value:.1f}")
    if value > 60:
        nth_percentile = p
        break

print(f"\nLowest percentile above 60 kcal: {nth_percentile}th")

  1-th percentile: 50.0
  2-th percentile: 50.0
  3-th percentile: 55.6
  4-th percentile: 70.0

Lowest percentile above 60 kcal: 4th


6. Хотя процентиль показывает нам, что у большинства конкурентов количество калорий намного выше, это неудобная концепция для использования в маркетинговых материалах.

Вместо этого давайте подсчитаем процент хлопьев, в которых содержится более 60 калорий на порцию. Сохраните свой ответ в переменной more_calories и распечатайте его

In [24]:
more_calories = np.mean(calorie_stats > 60) * 100

print(f"Is more calorific than 60 kcal: {more_calories:.2f} % marks")

Is more calorific than 60 kcal: 96.10 % marks


7. Это действительно высокий процент. Это будет очень полезно, когда мы будем продвигать CrunchieMunchies. Но один вопрос заключается в том, насколько велики различия в наборе данных? Можем ли мы сделать обобщение, что в большинстве злаков содержится около 100 калорий или разброс еще больше?

Рассчитайте величину отклонения, найдя стандартное отклонение, Сохраните свой ответ в calorie_std и распечатайте на терминале. Как мы можем включить эту ценность в наш анализ?

In [25]:
calorie_std = np.std(calorie_stats)

print(f"Standard deviation: {calorie_std:.2f} calories")
print(f"CrunchieMunchies are below average by {(average_calories - 60) / calorie_std:.2f} σ")

Standard deviation: 19.36 calories
CrunchieMunchies are below average by 2.42 σ


8. Напишите короткий абзац, в котором кратко изложите свои выводы и то, как, по вашему мнению, эти данные могут быть использованы в интересах Mycrunch при маркетинге CrunchieMunchies.

Of the 77 brands of competitors, 96% are more calorific than one serving of CrunchieMunchies. The average is 106.9 kcal, the median is 110 kcal: a typical serving is almost twice as high in calories as our 60. The spread is small (σ = 19 kcal), so it's not a matter of a pair of record holders about 100 kcal is the norm of the category.

Two formulations work best for marketing: "96% of the leading brands are more calorific than CrunchieMunchies" and "almost half as many calories as the average competitor." The percentiles are more accurate, but they need to be clarified — this is a losing proposition in advertising.

Calories are given per serving, and the serving sizes of the brands are different - this is not an apples-to-apples comparison; it would be more correct for 100 g. And low calorie content is not equal to "healthier": sugar, fiber, and sodium are missing from the set, so the protected statement is "fewer calories per serving" rather than "the healthiest choice."